src/hugr/view.rs

```diff
@@ -132,36 +183,33 @@ where
     }
 
     #[inline]
-    fn nodes(&self) -> Nodes<'_> {
+    fn nodes(&self) -> Self::Nodes<'_> {
         self.as_ref().graph.nodes_iter().map_into()
     }
 
     #[inline]
-    fn node_ports(&self, node: Node, dir: Direction) -> NodePorts {
+    fn node_ports(&self, node: Node, dir: Direction) -> Self::NodePorts<'_> {
         self.as_ref().graph.port_offsets(node.index, dir).map_into()
     }
 
     #[inline]
-    fn node_outputs(&self, node: Node) -> NodePorts {
-        self.as_ref().graph.output_offsets(node.index).map_into()
-    }
-
-    #[inline]
-    fn node_inputs(&self, node: Node) -> NodePorts {
-        self.as_ref().graph.input_offsets(node.index).map_into()
-    }
-
-    #[inline]
-    fn all_node_ports(&self, node: Node) -> NodePorts {
+    fn all_node_ports(&self, node: Node) -> Self::NodePorts<'_> {
         self.as_ref().graph.all_port_offsets(node.index).map_into()
     }
 
     #[inline]
-    fn linked_ports(&self, node: Node, port: Port) -> PortLinks<'_> {
+    fn linked_ports(&self, node: Node, port: Port) -> Self::PortLinks<'_> {
         let hugr = self.as_ref();
         let port = hugr.graph.port_index(node.index, port.offset).unwrap();
-        let links = hugr.graph.port_links(port);
-        PortLinks { hugr, links }
+        hugr.graph
+            .port_links(port)
+            .with_context(hugr)
+            .map_with_context(|(_, link), hugr| {
```

Comment:
Hmmm, alternatively, are we sure this shouldn't be a method on `hugr` rather than accessing the `hugr.graph` directly? Generally speaking I think the grand mixture of Node and Port vs NodeIndex and PortIndex that one sees when using a Hugr is a bit of a pain and it'd be nicer to standardize on Node/Port. Doing that everywhere is clearly beyond the scope of this PR but we could make a start....

Reply:
I don't think we want to expose anything handling `NodeIndex` or `SubportIndex` at the Hugr level.

Here we are wrapping the internal iterator from portgraph with works with `SubportIndex`, so we necessarily need some portgraph calls. (The context could have been a `&MultiPortgraph` instead of `&Hugr`, but it's the same here)